gusto/model.py

```diff
@@ -0,0 +1,228 @@
+from abc import ABCMeta, abstractproperty
+from collections import defaultdict
+from firedrake import SpatialCoordinate, sqrt, inner, interpolate, CellNormal, cross, Constant, as_vector
+from gusto.advection import SSPRK3, ThetaMethod
+from gusto.configuration import ShallowWaterParameters, CompressibleParameters
+from gusto.forcing import ShallowWaterForcing, CompressibleForcing
+from gusto.linear_solvers import ShallowWaterSolver, CompressibleSolver
+from gusto.state import State
+from gusto.transport_equation import VectorInvariant, AdvectionEquation, SUPGAdvection
+
+
+class Model(object, metaclass=ABCMeta):
+
+    eqn_opts = defaultdict(dict)
+
+    def __init__(self,
+                 mesh,
+                 timestepping=None,
+                 output=None,
+                 parameters=None,
+                 diagnostics=None,
+                 diagnostic_fields=None):
```

Comment:
OK, so a model creates a `State` object and then ties that together with a load of other stuff: advection schemes, diagnostic fields, etc...  The timestepper then uses the model to run the problem.  As commented below, I think with some judicious refactoring, there should be no need to inherit from `Model` for different model types.  Instead, one should be able to instantiate a model, poke properties of it into shape, and then hand that off to the Timestepper.

Then your `CompressibleEulerModel` and `ShallowWaterModel` can just be free functions that set up a `Model` instance appropriately.

One needs to be a little careful, because right now there's all sorts of action at a distance in the instantiation of a `Model`.  It makes `State`, but requires properties to be set in the instance that are not set in the constructor.  So either you are expecting that subclass constructors are called first (bad practice) or else that these are somehow class properties, rather than instance properties (it appears that the latter is the case).

I think this would make it also less confusing as to when the class instance has to provide instantiated objects (e.g. the parameters), and when it has to provide the constructor (e.g. `linear_solver`).